crates/red_knot_python_semantic/resources/mdtest/unary/instance.md

```diff
@@ -29,3 +31,82 @@ b = NoDunder()
 -b  # error: [unsupported-operator] "Unary operator `-` is unsupported for type `NoDunder`"
 ~b  # error: [unsupported-operator] "Unary operator `~` is unsupported for type `NoDunder`"
 ```
+
+## Not
+
+Not operator is inferred based on
+<https://docs.python.org/3/library/stdtypes.html#truth-value-testing>. An instance is True or False
+if the `__bool__` method says so.
+
+The `__len__` method on it's own will not determine the truthiness of an instance because `__bool__`
+method can override it.
+
+```py
+class AlwaysTrue:
+    def __bool__(self) -> Literal[True]:
+        return True
+
+# revealed: Literal[False]
+reveal_type(not AlwaysTrue())
+
+class AlwaysFalse:
+    def __bool__(self) -> Literal[False]:
+        return False
+
+# revealed: Literal[True]
+reveal_type(not AlwaysFalse())
+
+class BoolIsBool:
+    __bool__ = bool
+
+# revealed: bool
+reveal_type(not BoolIsBool())
+
+class NoBoolMethod: ...
```

Comment:
```suggestion
# At runtime, no `__bool__` and no `__len__` means truthy, but we can't rely on that, because
# a subclass could add a `__bool__` method.
class NoBoolMethod: ...
```